metricflow/model/semantics/linkable_spec_resolver.py

```diff
@@ -862,6 +899,39 @@ def _get_joined_elements(self, measure_semantic_model: SemanticModel) -> Linkabl
 
         return LinkableElementSet.merge_by_path_key((single_hop_elements, multi_hop_elements))
 
+    def get_all_linkable_metrics_for_semantic_model(self, starting_semantic_model: SemanticModel) -> LinkableElementSet:
+        """Get all metrics that can be joined to this semantic model using up to the max allowed join hops."""
+        linkable_metrics = LinkableElementSet()
+
+        semantic_models_and_join_paths: List[Tuple[SemanticModel, Optional[SemanticModelJoinPath]]] = [
+            (starting_semantic_model, None)
+        ]
+        for _ in range(self._max_entity_links):
+            next_semantic_models_and_join_paths: List[Tuple[SemanticModel, Optional[SemanticModelJoinPath]]] = []
+            for semantic_model, join_path in semantic_models_and_join_paths:
+                new_linkable_metrics = self.get_metrics_directly_joinable_to_semantic_model(
+                    semantic_model=semantic_model, using_join_path=join_path
+                )
+                linkable_metrics = LinkableElementSet.merge_by_path_key([linkable_metrics, new_linkable_metrics])
+
+                # Get next set of semantic models & join paths to iterate through.
+                for entity in semantic_model.entities:
+                    for next_semantic_model in self._entity_to_semantic_model[entity.name]:
+                        if next_semantic_model.name == semantic_model.name or (
+                            join_path and entity in join_path.entity_links
+                        ):
+                            continue
+                        join_path_elements = join_path.path_elements if join_path else ()
+                        new_join_path_element = SemanticModelJoinPathElement(
+                            semantic_model_reference=next_semantic_model.reference, join_on_entity=entity.reference
+                        )
+                        next_semantic_models_and_join_paths.append(
+                            (next_semantic_model, SemanticModelJoinPath(join_path_elements + (new_join_path_element,)))
+                        )
```

Comment:
Does this open the door to fanout on the intermediate step? Like I have a measure and that joins to the foreign `listing` entity via `account` in some other semantic model, and then we end up with `account__listing__some_listing_metric` but there's fanout between account and listing so our metrics aren't necessarily correct.

I don't see any checks for join validity within the entity block here.

Reply:
My brain realized this problem just when I was trying to fall asleep last night 🙃